#!/usr/bin/env python3"""This script creates a bed file for gene regions"""### ---------------------------------------- ###def get_genes(path):        bed_data = {'contig' : [],                'start' : [],                'end' : []}        with open(path) as gtf:                for g in gtf:                        g = g.replace('\n', '')                        if g.startswith('#'):                                continue                            elif g.split('\t')[2] == 'gene':                                contig, _, _, start, end, *_ = g.split('\t')                start, end = int(start), int(end)                                bed_data['contig'].append(contig)                bed_data['start'].append(start)                bed_data['end'].append(end)                        else:                                continue                    return pd.DataFrame(bed_data)### ---------------------------------------- ###def resolve_overlaps(raw_data):        clean_data = []        for contig in set(raw_data.contig):                # Paint contig regions from raw_data        contig_good_regions = np.zeros(raw_data.loc[raw_data.contig == contig, 'end'].max())                for _,(c,start,stop) in raw_data.loc[raw_data.contig == contig,].iterrows():                        contig_good_regions[start : stop] = 1                # Get bed intervals        end = -1        for n,s in enumerate(contig_good_regions):                        if s == 1 and end == -1:                                start = n                end = n                            elif s == 1 and end != -1:                                end = n                            elif s != 1 and end != -1:                                end += 1                clean_data.append(f'{contig}\t{start}\t{end}')                end = -1                            else:                                continue        return clean_data### ------------------MAIN------------------ ###import numpy as npimport pandas as pdfrom sys import argv### PARSE DATA ----------------------------- #### Get fasta pathgtf_path = argv[argv.index("--gtf") + 1]# Import file and get bed file for gene regionsbed_data = get_genes(gtf_path)# Resolving overlapsbed_data = resolve_overlaps(bed_data)bed_text = '\n'.join(bed_data)# Write to fileout_name = gtf_path.split('/')[-1].replace('.gtf', '') + '_genes.bed'with open(out_name, 'w') as out_file:        out_file.write(bed_text)